### ISSUE TITLE
Trained models for EGTEA Gaze+ dataset

### ISSUE THREAD

**Issue by ptirupat** [AllenXuuu/DCR#4]
Hello, 

I have trained your model on EGTEA_Gaze+ dataset using the following config file. The Top-5 Accuracy I get for this model trained on split1 with RGB features is 62.6% and when I train with Flow features, I get 39.2%. I tried a simple late fusion of these results which gives me a score of 54.7%. Can you please share your trained models and the script used to perform late fusion to reproduce your results?

#####################################################################
main_metric : 'T5_A'

eval:
  freq : 1
  data :
    name : 'EGTEA_GAZE+'
    split : 'valid1'
    
    feat_file : './datasets/EGTEA_GAZE+/features/TSN-C_3_egtea_action_CE_s1_rgb_model_best_fcfull_hd'    
    feature : 'TSN'
    feature_fps : 30
    feature_dim : 1024

    forward_frame : 8
    past_frame : 20
    fps : 4
    batch_size : 512
    num_workers : 2
    cache : false

train:
  data:
    name : 'EGTEA_GAZE+'
    split : 'train1'
    
    feat_file : './datasets/EGTEA_GAZE+/features/TSN-C_3_egtea_action_CE_s1_rgb_model_best_fcfull_hd'    
    feature : 'TSN'
    feature_fps : 30
    feature_dim : 1024

    forward_frame : 8
    past_frame : 20
    fps : 4
    batch_size : 512
    num_workers : 2
    cache : false

  optimizer:
    name: 'AdamW'
    base_lr : 5e-5
    betas : (0.9, 0.999)
    weight_decay : 1e-5
  
  max_epoch : 100
  scheduler:
    name : 'WarmupCos'
    warmup_epoch : 5
    step : [100]

curriculum:
  gamma_min : 0.95
  gamma_max : 1.
  
  
model:
  name : 'DCR'
  
  feat_dim: 1024
  past_frame : 20
  anticipation_frame : 4
  action_frame : 4
  
  reasoner:
    name : 'transformer'
    d_model : 1024
    nhead : 16
    dff : 4096
    depth : 6
    dropout : 0.1
    pe_type : 'learnable'
    
  classifier:
    dropout: 0.4
    action: true
    verb: true
    noun : true

  loss:
    name : 'CE'
    verb : 0.5
    noun : 0.5
    feat_mse : 1.
    next_cls : 0.5
    smooth: 0.4
#####################################################################

**Comment by AllenXuuu** [comment]
Thanks for your interests. We release trained models, predicted logits and fused results on EGTEA Gaze+ at
https://drive.google.com/drive/folders/1zZ3NShwEYcU27gjjG_3aFRYMBs1Yia30?usp=sharing.


The fusion on EGTEA Gaze+ uses class-wised picked weights considering different roles of two modals on different classes.


import pickle as pkl
import os
import numpy as np

val1_flow = 'EG_SP1_FLOW.pt.valid1.pkl'
val2_flow = 'EG_SP2_FLOW.pt.valid2.pkl'
val3_flow = 'EG_SP3_FLOW.pt.valid3.pkl'
val1_rgb  = 'EG_SP1_RGB.pt.valid1.pkl'
val2_rgb  = 'EG_SP2_RGB.pt.valid2.pkl'
val3_rgb  = 'EG_SP3_RGB.pt.valid3.pkl'

weight = {}
weight['rgb']  = [0.683, 0.991, 0.719, 1.   , 1.   , 0.972, 0.796, 1.   , 1.   ,1.   , 0.55 , 1.   , 0.589, 1.   , 0.652, 0.778, 0.801, 0.77 ,0.895, 0.852, 1.   , 0.939, 1.   , 0.812, 0.942, 0.958, 1.   ,0.786, 0.926, 1.   , 0.942, 1.   , 1.   , 0.928, 1.   , 1.   ,1.   , 1.   , 0.674, 1.   , 1.   , 0.864, 1.   , 0.969, 1.   ,1.   , 1.   , 0.792, 0.695, 0.781, 1.   , 0.858, 0.741, 0.527,1.   , 1.   , 0.996, 0.816, 0.976, 0.743, 0.977, 0.886, 1.   ,1.   , 0.704, 0.764, 0.478, 1.   , 0.834, 1.   , 1.   , 1.   ,0.975, 1.   , 1.   , 1.   , 1.   , 0.934, 1.   , 1.   , 1.   ,1.   , 1.   , 0.678, 1.   , 1.   , 1.   , 1.   , 1.   , 1.   ,0.679, 1.   , 1.   , 1.   , 1.   , 1.   , 1.   , 1.   , 1.   ,0.1  , 0.122, 0.977, 0.77 , 1.   , 0.887, 0.796]
weight['flow'] = [0.419, 0.   , 0.159, 0.038, 0.239, 0.   , 0.113, 0.   , 0.231,0.335, 0.192, 0.   , 0.084, 0.009, 0.123, 0.   , 0.31 , 0.031,0.   , 0.074, 0.   , 0.224, 0.205, 0.197, 0.195, 0.066, 0.27 ,0.   , 0.   , 0.   , 0.153, 0.   , 0.   , 0.   , 0.   , 0.162,0.033, 0.007, 0.086, 0.   , 0.183, 0.   , 0.023, 0.   , 0.492,0.   , 0.   , 0.211, 0.   , 0.018, 0.   , 0.   , 0.043, 0.265,0.471, 0.   , 0.225, 0.24 , 0.   , 0.318, 0.   , 0.03 , 0.211,0.   , 0.136, 0.3  , 0.457, 0.   , 0.   , 0.045, 0.   , 0.055,0.   , 0.   , 0.   , 0.154, 0.166, 0.512, 0.215, 0.126, 0.345,0.   , 0.   , 0.   , 0.   , 0.   , 0.088, 0.   , 0.   , 0.893,0.222, 0.   , 0.   , 0.   , 0.   , 0.399, 0.214, 0.136, 0.007,0.   , 0.953, 0.   , 0.   , 0.61 , 0.286, 0.133]

val1_flow = pkl.load(open(val1_flow,'rb'))
val2_flow = pkl.load(open(val2_flow,'rb'))
val3_flow = pkl.load(open(val3_flow,'rb'))
val1_rgb = pkl.load(open(val1_rgb,'rb'))
val2_rgb = pkl.load(open(val2_rgb,'rb'))
val3_rgb = pkl.load(open(val3_rgb,'rb'))

gt1 = val1_rgb['gt']
gt2 = val2_rgb['gt']
gt3 = val3_rgb['gt']
val1_flow = val1_flow['pred']['action']
val2_flow = val2_flow['pred']['action']
val3_flow = val3_flow['pred']['action']
val1_rgb = val1_rgb['pred']['action']
val2_rgb = val2_rgb['pred']['action']
val3_rgb = val3_rgb['pred']['action']

val1_fuse = val1_flow * weight['flow'] + val1_rgb * weight['rgb']
val2_fuse = val2_flow * weight['flow'] + val2_rgb * weight['rgb']
val3_fuse = val3_flow * weight['flow'] + val3_rgb * weight['rgb']

def calc_t5_acc_rec(pred,gt):
    rank = np.argsort(pred,1)[:,::-1][:,:5]
    is_correct = np.sum(rank == np.expand_dims(gt,1),1)
    acc = np.mean(is_correct)
    
    classes = np.unique(gt)
    recalls = []
    for c in classes:
        is_correct_this_class = is_correct[gt == c]
        recalls.append(is_correct_this_class.mean())
    rec = np.nanmean(recalls)
    return acc,rec


acc1,rec1 = calc_t5_acc_rec(val1_fuse, gt1['action'])
acc2,rec2 = calc_t5_acc_rec(val2_fuse, gt2['action'])
acc3,rec3 = calc_t5_acc_rec(val3_fuse, gt3['action'])


acc_avg = (acc1+acc2+acc3) / 3
rec_avg = (rec1+rec2+rec3) / 3

print('Split 1: T5-ACC: %.3f, RECALL@5: %.3f' % (acc1,rec1))
print('Split 2: T5-ACC: %.3f, RECALL@5: %.3f' % (acc2,rec2))
print('Split 3: T5-ACC: %.3f, RECALL@5: %.3f' % (acc3,rec3))
print('AVG    : T5-ACC: %.3f, RECALL@5: %.3f' % (acc_avg,rec_avg))

with open('EG_SP1_fusion.pkl','wb') as f:
    pkl.dump({
        'gt' : gt1, 'pred' : val1_fuse
    },f)
    
with open('EG_SP2_fusion.pkl','wb') as f:
    pkl.dump({
        'gt' : gt2, 'pred' : val2_fuse
    },f)
    
with open('EG_SP3_fusion.pkl','wb') as f:
    pkl.dump({
        'gt' : gt2, 'pred' : val3_fuse
    },f)

**Comment by ptirupat** [comment]
Thank you for your quick response and for sharing the trained models. Can you please explain how these class-specific weights were computed?

**Comment by AllenXuuu** [comment]
selected via cross validating

**Comment by ptirupat** [comment]
Sorry, I did not understand. Do you perform some sort of grid search to find the best weights for all the classes?

**Comment by AllenXuuu** [comment]
Weights are additionally trained, as to learn a weighted sum from two modals to fusing. We also release stronger models without class-wise fusion now.
